src/Streamly/Internal/Data/Parse.hs

```diff
@@ -0,0 +1,254 @@
+-- Terminating folds
+-------------------------------------------------------------------------------
+--
+-- XXX any/all should be terminating folds when we have a terminating fold
+-- type.
+--
+-- >>> S.parse (PR.any (== 0)) $ S.fromList [1,0,1]
+-- > Right True
+--
+{-# INLINABLE any #-}
+any :: Monad m => (a -> Bool) -> Parse m a Bool
+any predicate = Parse step initial (\x -> return $ Right (0,x))
+    where
+    initial = return False
+    step s a = return $
+        if s
+        then Halt True
+        else
+            if predicate a
+            then Halt True
+            else Keep 0 False
+
+-- >>> S.parse (PR.any (== 0)) $ S.fromList [1,0,1]
+-- > Right False
+--
+{-# INLINABLE all #-}
+all :: Monad m => (a -> Bool) -> Parse m a Bool
+all predicate = Parse step initial (\x -> return $ Right (0,x))
+    where
+    initial = return True
+    step s a = return $
+        if s
+        then
+            if predicate a
+            then Keep 0 True
+            else Halt False
+        else Halt False
+
+-------------------------------------------------------------------------------
+-- Taking elements
+-------------------------------------------------------------------------------
+--
+-- | Stops after taking exactly @n@ input elements.
+--
+-- * Stops - after @n@ elements.
+-- * Fails - if the stream ends before it can collect @n@ elements.
+--
+-- >>> S.parse (PR.takeExact 4 FL.toList) $ S.fromList [1,0,1]
+-- > Left "takeExact: Expecting exactly 4 elements, got 3"
+--
+-- /Internal/
+--
+{-# INLINABLE takeExact #-}
+takeExact :: Monad m => Int -> Fold m a b -> Parse m a b
+takeExact n (Fold fstep finitial fextract) = Parse step initial extract
+
+    where
+
+    initial = (Tuple' 0) <$> finitial
+
+    step (Tuple' i r) a = do
+        res <- fstep r a
+        let i1 = i + 1
+            s1 = Tuple' i1 res
+        return $ if i1 < n then Hold s1 else Halt s1
+
+    extract (Tuple' i r) = fmap f (fextract r)
+
+        where
+
+        err =
+               "takeExact: Expecting exactly " ++ show n
+            ++ " elements, got " ++ show i
+
+        f x =
+            if n == i
+            then Right (0, x)
+            else Left err
+
+-- | Take at least @n@ input elements, but can collect more.
+--
+-- * Stops - never.
+-- * Fails - if the stream end before producing @n@ elements.
+--
+-- >>> S.parse (PR.takeAtLeast 4 FL.toList) $ S.fromList [1,0,1]
+-- > Left "takeAtLeast: Expecting at least 4 elements, got only 3"
+--
+-- >>> S.parse (PR.takeAtLeast 4 FL.toList) $ S.fromList [1,0,1,0,1]
+-- > Right [1,0,1,0,1]
+--
+-- /Internal/
+--
+{-# INLINABLE takeAtLeast #-}
+takeAtLeast :: Monad m => Int -> Fold m a b -> Parse m a b
+takeAtLeast n (Fold fstep finitial fextract) = Parse step initial extract
+
+    where
+
+    initial = (Tuple' 0) <$> finitial
+
+    step (Tuple' i r) a = do
+        res <- fstep r a
+        let i1 = i + 1
+            s1 = Tuple' i1 res
+        return $ Hold s1
```

Comment:
This `step` makes more sense in my opinion. 
```
    step (Tuple' i r) a = do
        res <- fstep r a
        let i1 = i + 1
            s1 = Tuple' i1 res
        if i1 < n
        then return $ Hold s1
        else return $ Keep 0 s1
```
I'm also curious how `takeAtleast 5 <|> takeExact 10` would work with different implementations of `step`.

Reply:
In my opinion alternative instance should be as greedy as possible.

Reply:
Done.